pkg/gateway/controller.go

```diff
@@ -0,0 +1,217 @@
+	c.store, c.informer = cache.NewInformer(&cache.ListWatch{
+		ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+			return gwClientset.List(metav1.ListOptions{})
+		},
+		WatchFunc: gwClientset.Watch,
+	}, &unstructured.Unstructured{}, 0, cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			key, err := cache.MetaNamespaceKeyFunc(obj)
+			if err == nil {
+				c.queue.Add(key)
+			}
+		},
+		UpdateFunc: func(obj interface{}, new interface{}) {
+			key, err := cache.MetaNamespaceKeyFunc(new)
+			klog.V(2).Infof("GatewayStatus %q updated", key)
+			if err == nil {
+				c.queue.Add(key)
+			}
+		},
+		DeleteFunc: func(obj interface{}) {
+			key, _ := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+			klog.V(2).Infof("GatewayStatus %q deleted", key)
+		},
+	})
+	go c.informer.Run(c.stopCh)
+	go c.runWorker()
+
+	return nil
+}
+
+func (c *Controller) Stop() {
+	close(c.stopCh)
+	c.queue.ShutDown()
+	klog.Infof("Gateway status Controller stopped")
+}
+
+func (c *Controller) runWorker() {
+	for {
+		keyObj, shutdown := c.queue.Get()
+		if shutdown {
+			klog.Infof("Lighthouse watcher for Gateways stopped")
+			return
+		}
+
+		key := keyObj.(string)
+		func() {
+			defer c.queue.Done(key)
+			obj, exists, err := c.store.GetByKey(key)
+			if err != nil {
+				klog.Errorf("Error retrieving gateway with key %q from the cache: %v", key, err)
+				// requeue the item to work on later
+				c.queue.AddRateLimited(key)
+				return
+			}
+			if exists {
+				c.gatewayCreatedOrUpdated(obj.(*unstructured.Unstructured))
+			}
+			c.queue.Forget(key)
+		}()
+	}
+}
+
+func (c *Controller) gatewayCreatedOrUpdated(obj *unstructured.Unstructured) {
+	haStatus, connections, ok := getGatewayStatus(obj)
+	if !ok || haStatus != "active" {
+		return
+	}
+	var newMap map[string]bool
+	currentMap := c.getClusterStatusMap()
+	for _, connection := range connections {
+		connectionMap := connection.(map[string]interface{})
+
+		status, found, err := unstructured.NestedString(connectionMap, "status")
+		if err != nil || !found {
+			klog.Errorf("status field not found in %#v", connectionMap)
+		}
+		clusterId, found, err := unstructured.NestedString(connectionMap, "endpoint", "cluster_id")
+		if !found || err != nil {
+			klog.Errorf("clusterId field not found in %#v", connectionMap)
+			return
+		}
+
+		if status == "connected" {
+			_, found := currentMap[clusterId]
+			if !found {
+				if newMap == nil {
+					newMap = copyMap(currentMap)
+				}
+				newMap[clusterId] = true
+			}
+		} else {
+			_, found = currentMap[clusterId]
+			if found {
+				if newMap == nil {
+					newMap = copyMap(currentMap)
+				}
+				delete(newMap, clusterId)
+			}
+		}
+	}
+
+	if newMap != nil {
+		klog.Infof("Updating the gateway status %#v ", newMap)
+		c.clusterStatusMap.Store(newMap)
+	}
+}
+
+func getGatewayStatus(obj *unstructured.Unstructured) (haStatus string, connections []interface{}, gwStatus bool) {
```

Comment:
I wonder why the inter is forcing the use of named params here - it's not considered best practice in most cases.  `gwStatus` is unused - I'm surprised it didn't flag that 😄  Anyway just an observation - it's fine either way.